File: 3. Behavioral Design Patterns/Observer/C#/ObserverTemplate.cs
// Observer Pattern – C# Template

using System.Collections.Generic;

namespace ObserverTemplate
{
    interface IObserver { void Update(); }
    interface ISubject { void Attach(IObserver o); void Detach(IObserver o); void Notify(); }

    class ConcreteSubject : ISubject
    {
        private readonly List<IObserver> _observers = new();
        public void Attach(IObserver o) => _observers.Add(o);
        public void Detach(IObserver o) => _observers.Remove(o);
        public void Notify() { foreach (var o in _observers) o.Update(); }
        public void DoSomething() { /* change state */ Notify(); }
    }

    class ConcreteObserver : IObserver { public void Update() { /* react to change */ } }

    class Program { static void Main(){ var s = new ConcreteSubject(); s.Attach(new ConcreteObserver()); s.DoSomething(); } }
}

File: 3. Behavioral Design Patterns/Observer/Kotlin/ObserverTemplate.kt
// Observer Pattern – Kotlin Template

interface Observer { fun update() }
interface Subject { fun attach(o: Observer); fun detach(o: Observer); fun notify() }

class ConcreteSubject : Subject {
    private val observers = mutableListOf<Observer>()
    override fun attach(o: Observer) = observers.add(o).let{}
    override fun detach(o: Observer) = observers.remove(o).let{}
    override fun notify() = observers.forEach { it.update() }
    fun doSomething() { /* change state */ notify() }
}

class ConcreteObserver : Observer { override fun update() { /* react to change */ } }

fun main(){ val s = ConcreteSubject(); s.attach(ConcreteObserver()); s.doSomething() }
